src/app/Http/Traits/SignActionDocumentSignatureTrait.php

```diff
@@ -2,120 +2,404 @@
+        }
+
+        $signature = $this->doSignature($setupConfig, $data, $documentSignatureEsignData);
+        return $signature;
+    }
+
+    /**
+     * doSignature
+     *
+     * @param  array $setupConfig
      * @param  collection $data
+     * @param  array $documentSignatureEsignData
      * @return collection
      */
-    protected function findNextDocumentSent($data)
+    protected function doSignature($setupConfig, $data, $documentSignatureEsignData)
     {
-        $nextDocumentSent = DocumentSignatureSent::where('ttd_id', $data->ttd_id)
-                                                    ->where('urutan', $data->urutan + 1)
-                                                    ->first();
-        return $nextDocumentSent;
+        $documentData = ($documentSignatureEsignData['isSignedSelf'] == true) ? $data : $data->documentSignature;
+        $setNewFileData = $this->setNewFileData($documentData);
+        $pdfFile = $this->pdfFile($documentData, $setNewFileData['verifyCode'], $documentSignatureEsignData);
+        $response = Http::withHeaders([
+            'Authorization' => 'Basic ' . $setupConfig['auth'], 'Cookie' => 'JSESSIONID=' . $setupConfig['cookies'],
+        ])->attach('file', $pdfFile, $documentData->file)->post($setupConfig['url'] . '/api/sign/pdf', [
+            'nik'           => $setupConfig['nik'],
+            'passphrase'    => $documentSignatureEsignData['passphrase'],
+            'tampilan'      => 'invisible',
+            'image'         => 'false',
+        ]);
+        if ($response->status() != Response::HTTP_OK) {
+            $bodyResponse = json_decode($response->body());
+            $this->setPassphraseSessionLog($response, $documentData, SignatureDocumentTypeEnum::UPLOAD_DOCUMENT(), $documentSignatureEsignData);
+            $logData = [
+                'message' => 'Gagal melakukan tanda tangan elektronik',
+                'longMessage' => $bodyResponse->error
+            ];
+            // Set return failure esign
+            return $this->esignFailedExceptionResponse($logData, $documentSignatureEsignData, $data->id, SignatureDocumentTypeEnum::UPLOAD_DOCUMENT());
+        } else {
+            //Save new file & update status
+            $doUpdate = $this->saveNewFile($response, $data, $setNewFileData, $documentSignatureEsignData);
+            $this->setPassphraseSessionLog($response, $data, SignatureDocumentTypeEnum::UPLOAD_DOCUMENT(), $documentSignatureEsignData);
+            $this->checkIsLastItemQueueRedis($data->id, $documentSignatureEsignData);
+            return $doUpdate;
+        }
+    }
+
+    protected function checkIsLastItemQueueRedis($id, $documentSignatureEsignData)
+    {
+        if (
+            $documentSignatureEsignData['medium'] == MediumTypeEnum::WEBSITE() &&
+            $documentSignatureEsignData['esignMethod'] == SignatureMethodTypeEnum::MULTIFILE() &&
+            $id == end($documentSignatureEsignData['items']))
+            {
```

Comment:
I actually was thinking something like this:

```suggestion
if (
            $documentSignatureEsignData['medium'] == MediumTypeEnum::WEBSITE() &&
            $documentSignatureEsignData['esignMethod'] == SignatureMethodTypeEnum::MULTIFILE() &&
            $id == end($documentSignatureEsignData['items'])
) {
```

personally I think this style of formatting is easier to reach, but feel free to ignore it if it felt too picky 🤣

Reply:
hmmmm i agree with it, changes will be add later

Reply:
fix here mas [fix: if statement](https://github.com/jabardigitalservice/office-services/pull/583/commits/ea399a1ce966ae4caa97d8f24a0d4e9476865169)